src/core/ddsi/src/ddsi_typelib.c

```diff
@@ -919,4 +1123,70 @@ void ddsi_type_pair_free (struct ddsi_type_pair *type_pair)
   ddsrt_free (type_pair);
 }
 
+dds_return_t ddsi_wait_for_type_resolved (struct ddsi_domaingv *gv, const ddsi_typeid_t *type_id, dds_duration_t timeout, struct ddsi_type **type, ddsi_type_include_deps_t resolved_kind, ddsi_type_request_t request)
```

Comment:
Something that occurred to me just now that slipped through the cracks some time ago (and that would be best dealt with an issue and some future PR) is that some change is needed to interrupt this when trying to delete any entity that was pinned while calling this function or when shutting down cyclone. It is not difficult, just a bit of bookkeeping and returning an error.